linode/terraform-provider-linode
# new: Add CustomDiff and PlanModifier for case-insensitive sets

## 📝 Description

This change alters all Framework and SDKv2 resources with tags to ignore the case during diffs on the `tags` field. This is achieved through the new `CaseInsensitiveSet` CustomDiff and PlanModifier added in this PR.

These work by modifying the plan to match values in the state regardless if they are equivalent ignoring case. This allows `terraform plan` to accurately report changes certain values in the set while not reporting case changes.

Additionally, this PR resolves an issue that prevented tags from being updated to an empty set in the `linode_domain`, `linode_instance`, and `linode_lke_cluster` resources.

This PR also adds notes to the documentation making it clear that tags are fully case-insensitive.

## ✔️ How to Test

The following test steps assume you have pulled this PR down locally.

### Unit Testing

```bash
make unit-test
```

### Integration Testing

```bash
make int-test PKG_NAME=linode/domain
make int-test PKG_NAME=linode/firewall
make int-test PKG_NAME=linode/instance
make int-test PKG_NAME=linode/lke
make int-test PKG_NAME=linode/nb
make int-test PKG_NAME=linode/volume
```


### Manual Testing

1. In a Terraform provider sandbox (e.g. dx-devenv), apply the following configuration:

```terraform
# ...

locals {
  tags = ["foo", "bar"]
}

resource "linode_instance" "test" {
  label = "test"
  region = "us-mia"
  type = "g6-nanode-1"

  tags = local.tags
}

resource "linode_volume" "test" {
  label = "test"
  region = "us-mia"
  size = 20
  tags = local.tags
}

resource "linode_lke_cluster" "test" {
  label = "test"
  region = "us-mia"
  tags = local.tags
  k8s_version = "1.28"

  pool {
    type = "g6-standard-1"
    count = 1
  }
}

resource "linode_domain" "test" {
  domain = "${random_string.domain_name.result}.com"
  soa_email = "test@${random_string.domain_name.result}.com"
  type = "master"
  tags = local.tags
}

resource "random_string" "domain_name" {
  length = 32
  special = false
  numeric = false
}

resource "linode_firewall" "test" {
  label = "test"
  tags = local.tags

  inbound_policy = "DROP"
  outbound_policy = "DROP"
}

resource "linode_nodebalancer" "test" {
  region = "us-mia"
  label = "test"
  tags = local.tags
}
```

2. Observe the configuration applies successfully.
3. Update the `tags` local to change the case without any other changes:

```terraform
locals {
  tags = ["fOO", "Bar"]
}
```
4. Apply the configuration and ensure no changes were planned/made.
5. Update the `tags` local to alter the tags:

```terraform
locals {
  tags = ["wow", "bar"]
}
```

6. Apply the configuration and ensure the new tags are reflected in the plan.
7. Apply the configuration again and ensure no changes were planned/made.
